Userland/Libraries/LibWeb/DOM/Document.cpp

```diff
@@ -3489,4 +3489,37 @@ Painting::ViewportPaintable* Document::paintable()
     return static_cast<Painting::ViewportPaintable*>(Node::paintable());
 }
 
+// https://html.spec.whatwg.org/multipage/browsing-the-web.html#update-document-for-history-step-application
+void Document::update_for_history_step_application(JS::NonnullGCPtr<HTML::SessionHistoryEntry> entry, bool do_not_reactive, size_t script_history_length, size_t script_history_index)
+{
+    // 1. Let documentIsNew be true if document's latest entry is null; otherwise false.
+    auto document_is_new = !m_latest_entry;
+
+    // 2. Let documentsEntryChanged be true if document's latest entry is not entry; otherwise false.
+    auto documents_entry_changed = m_latest_entry != entry;
+
+    // 3. Set document's history object's index to scriptHistoryIndex.
+    history()->m_index = script_history_index;
+
+    // 4. Set document's history object's length to scriptHistoryLength.
+    history()->m_length = script_history_length;
+
+    // 5. If documentsEntryChanged is true, then:
+    if (documents_entry_changed) {
+        // FIXME: Implement this.
+    }
+
+    // 6. If documentIsNew is true, then:
+    if (document_is_new) {
+        // FIXME: 1. Try to scroll to the fragment for document.
+        // FIXME: 2. At this point scripts may run for the newly-created document document.
```

Comment:
Was there some other spec concept here that we wait for in the parser in  order to run scripts? Or is that all still mega ad-hoc

Reply:
yes it seems so https://html.spec.whatwg.org/multipage/document-lifecycle.html#read-html:scripts-may-run-for-the-newly-created-document
let's take care of it in a separate PR? :)

Reply:
I was thinking of step 5 here in the end:
<https://html.spec.whatwg.org/multipage/parsing.html#the-end>

But it seems like that new concept is just kinda... shoehorned in? Not sure how it fits into "is ready to be parser executed" and such. So yeah, separate PR :)